File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        bracketMap = {
            "(":")",
            "[":"]",
            "{":"}"
        }

        for char in s:
            if char in bracketMap:
                stack.append(bracketMap[char])
            else:
                if not stack or stack[-1] != char:
                    return False
                stack.pop()
                
        return True if not stack else False

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
    
        unordered_map<char, char> bracketMap {
            {'(', ')'},
            {'[', ']'},
            {'{', '}'} 
        };

        stack<char> bracketStack;

        for (const char &c : s) {
            if (bracketMap.contains(c)) {
                bracketStack.push(c);
            }
            else {
                if (bracketStack.empty()) {
                    return false;
                }
                if (bracketMap[bracketStack.top()] != c) {
                    return false;
                }
                bracketStack.pop();
            }
        }

        return bracketStack.empty();
    }
};